Add fee_est tool for debugging fee estimation code

ryanofsky opened the pull request:
This PR adds an `-estlog` option for saving live fee estimation data from a bitcoin node, and a `fee_est` command line tool for processing the data and testing fee estimation code.

The idea is to make it easier to test improvements to fee estimation like https://github.com/bitcoin/bitcoin/pull/10199 in a more systematic and reproducible way.

Some documentation is in [`src/test/fee_est/README.md`](https://github.com/ryanofsky/bitcoin/blob/pr/fee/src/test/fee_est/README.md)

Sample log file: [`est.log.xz`](https://storage.googleapis.com/ryanofsky/est.log.xz) (65M)

Sample [`fee_est.cpp`](https://github.com/ryanofsky/bitcoin/blob/pr/fee/src/test/fee_est/fee_est.cpp)  graph output: 

![graph](https://cloud.githubusercontent.com/assets/7133040/26331057/f2de8d0c-3f1c-11e7-99ea-de1801d13b05.png)

ryanofsky committed:
Add -estlog option for saving live fee estimation data

ryanofsky committed:
Add fee_est tool for debugging fee estimation code

paveljanik commented:
This is interesting but still no comment.

Concept ACK

I'll play around with this a bit.

```
Wshadow statistics: 
   1 policy/fees_input.cpp:127:23: warning: declaration shadows a local variable [-Wshadow]
   1 policy/fees_input.cpp:197:38: warning: declaration shadows a local variable [-Wshadow]
   1 policy/fees_input.cpp:229:22: warning: declaration shadows a local variable [-Wshadow]
   1 policy/fees_input.cpp:233:27: warning: declaration shadows a local variable [-Wshadow]
   1 policy/fees_input.cpp:27:61: warning: declaration shadows a field of 'CBlockPolicyInput' [-Wshadow]
   1 test/fee_est/fee_est.cpp:301:34: warning: declaration shadows a local variable [-Wshadow]
   1 test/fee_est/fee_est.cpp:79:20: warning: declaration shadows a field of 'TxData' [-Wshadow]
   1 test/fee_est/fee_est.cpp:79:29: warning: declaration shadows a field of 'TxData' [-Wshadow]
   1 test/fee_est/fee_est.cpp:79:39: warning: declaration shadows a field of 'TxData' [-Wshadow]
```

paveljanik commented:
When estimate log doesn't have enough data, `fee_est` is generating empty data in the HTML file. Emit some warning in such cases?

paveljanik commented:
For perfect output, there should be 3rd dimension - time or current block height when the tx was first seen :-)

jnewbery commented:
Concept ACK. I think this could be a useful tool.

Could this be split into two PRs to aid reviewers? The first PR would cover the `-estlog` option and writing the fee estimation data to disk. The second PR would be for a tool to read and graph the logs.

Parsing and graphing json files seems like a problem that has probably been solved many times before. If I was approaching this, I'd look at implementing this as a script in /contrib or a separate repository. Was there a particular reason you chose to implement this as a new C++ program?

laanwj commented:
> Could this be split into two PRs to aid reviewers? The first PR would cover the -estlog option and writing the fee estimation data to disk. The second PR would be for a tool to read and graph the logs.

I think this is useful.

However I'm not sure the analysis tool belongs in this repository. As it's specific to developers debugging the fee estimation code it's not something we want to ship with the distribution, or install by default. One place it could be is e.g. https://github.com/bitcoin-core/bitcoin-maintainer-tools . You could still document or refer to it from here.

ryanofsky commented:
Agree that fee_est tool shouldn't be installed, since it's a tool specifically for made for modifying and debugging fee estimation code.

But it would be awkward to use and maintain from a separate repository because it links and calls into the fee estimation code. (The tool works by piping historical data into the fee estimator so it's possible to make experimental changes to fee estimation and see how those changes affect output and internal state.)

As far as build / distribution is concerned I think it makes sense to think of it more like a unit test or benchmark than a maintainer tool.

laanwj commented:
> But it would be awkward to use and maintain from a separate repository because it links and calls into the fee estimation code. 

Yes, I agree, that's unfortunately true.

> Parsing and graphing json files seems like a problem that has probably been solved many times before. If I was approaching this, I'd look at implementing this as a script in /contrib or a separate repository.

I agree - especially with the references to cloudflare CDN and such. It feels a bit ugly to have that in the C++ code. Better to spit out e.g. JSON or CSV or whatever is convenient, then use a separate script for pretty formatting. This would also keep open the option of using different (non-web) visualization tools.

practicalswift commented on src/test/fee_est/README.md:
```diff
@@ -0,0 +1,37 @@
+estimation code using transaction data gathered from live bitcoin nodes.
+
+Transaction data can be collected by running bitcoind with `-estlog` parameter
+which will produce newline-delimited json file (http://ndjson.org/,
```
````
2018-09-11 16:31:38 mdl(pr=10443): src/test/fee_est/README.md:8: MD034 Bare URL used
2018-09-11 16:31:38 mdl(pr=10443): src/test/fee_est/README.md:9: MD034 Bare URL used
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+        } else if (sample.size() < SAMPLE_TXS) {
+            sample.emplace_back(&entry);
+        } else {
+            int j = rand() % i;
```
Mod by zero here.

practicalswift replied:
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:177:21: warning: rand() has limited randomness; use C++11 random library instead [cert-msc30-c]
```

practicalswift commented on src/policy/fees.h:
```diff
@@ -190,11 +191,12 @@ class CBlockPolicyEstimator
 
     /** Process a transaction accepted to the mempool*/
-    void processTransaction(const CTxMemPoolEntry& entry, bool validFeeEstimate);
+    void processTx(const uint256& hash, unsigned int height, CAmount fee, size_t size, Block* block, bool valid);
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees.h:199:10: warning: function 'CBlockPolicyEstimator::processTx' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name]
```

practicalswift commented on src/policy/fees.cpp:
```diff
@@ -9,11 +9,17 @@
 static constexpr double INF_FEERATE = 1e99;
 
+struct CBlockPolicyEstimator::Block {
+    Block(int height) : height(height) {}
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees.cpp:17:5: warning: single-argument constructors must be marked explicit to avoid unintentional implicit conversions [google-explicit-constructor]
```

practicalswift replied:
```
2018-09-20 05:13:44 cppcheck(pr=10443): [src/policy/fees.cpp:17]: (style) Struct 'Block' has a constructor with 1 argument that is not explicit.
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+    const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
+    CBlockPolicyEstimator& estimator,
+    CTxMemPool& pool)
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:149:17: warning: parameter 'pool' is unused [misc-unused-parameters]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+            fee.pushKV("blocks", int(target));
+            int64_t feeRate = estimator.estimateSmartFee(target, nullptr /* confTarget */, conservative).GetFeePerK();
+            fee.pushKV("feeRate", feeRate);
+            fees.push_back(std::move(fee));
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:163:28: warning: passing result of std::move() as a const reference argument; no move will actually happen [hicpp-move-const-arg]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+        ++i;
+    }
+
+    if (sample.size() > 0) {
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:185:9: warning: the 'empty' method should be used to check for emptiness instead of 'size' [readability-container-size-empty]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+                tx.pushKV("feeRate", feeRate);
+                tx.pushKV("blocks", entry->second.actualBlocks);
+                tx.pushKV("height", entry->second.height);
+                txs.push_back(std::move(tx));
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:205:31: warning: passing result of std::move() as a const reference argument; no move will actually happen [hicpp-move-const-arg]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+}
+
+void PrintCross(const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:229:38: warning: parameter 'targets' is unused [misc-unused-parameters]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+
+void PrintCross(const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
+    CBlockPolicyEstimator& estimator,
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:230:28: warning: parameter 'estimator' is unused [misc-unused-parameters]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+void PrintCross(const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
+    CBlockPolicyEstimator& estimator,
+    CTxMemPool& pool)
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/test/fee_est/fee_est.cpp:231:17: warning: parameter 'pool' is unused [misc-unused-parameters]
```

practicalswift commented on src/policy/fees_input.h:
```diff
@@ -0,0 +1,55 @@
+    void processTx(const uint256& hash, unsigned int height, CAmount fee, size_t size, Block* block, bool valid);
+
+    /** Remove a transaction from the mempool tracking stats */
+    void removeTx(const uint256& hash, bool in_block);
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.h:33:10: warning: function 'FeeEstInput::removeTx' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name]
```

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,218 @@
+        tx.pushKV("fee", fee);
+        tx.pushKV("size", uint64_t(size));
+        if (block) {
+            block->json_txs.push_back(std::move(tx));
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.cpp:30:39: warning: passing result of std::move() as a const reference argument; no move will actually happen [hicpp-move-const-arg]
```

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,218 @@
+            block->json_txs.push_back(std::move(tx));
+        } else {
+            UniValue value(UniValue::VOBJ);
+            value.pushKV("tx", std::move(tx));
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.cpp:33:32: warning: passing result of std::move() as a const reference argument; no move will actually happen [hicpp-move-const-arg]
```

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,218 @@
+    if (filename.empty()) {
+        log.reset();
+    } else {
+        log.reset(new std::ofstream(filename, std::ofstream::out | std::ofstream::app));
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.cpp:126:13: warning: use std::make_unique instead [modernize-make-unique]
```

`MakeUnique` in our case :-)

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,218 @@
+    return true;
+}
+
+bool FeeEstInput::readLog(const std::string& filename, std::function<bool(UniValue&)> filter)
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.cpp:143:87: warning: the parameter 'filter' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
```

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,218 @@
+        const UniValue& read = value["read"];
+        if (read.isStr()) {
+            std::vector<unsigned char> data = ParseHex(read.get_str());
+            unsigned short randv = 0;
```
```
2018-09-20 04:48:56 clang-tidy(pr=10443): src/policy/fees_input.cpp:198:13: warning: consider replacing 'unsigned short' with 'uint16' [google-runtime-int]
```

practicalswift replied:
```
2018-09-22 22:51:38 cpplint(pr=10443): src/policy/fees_input.cpp:198:  Use int16/int64/etc, rather than the C type short  [runtime/int] [4]
```

practicalswift commented on src/policy/fees_input.h:
```diff
@@ -0,0 +1,55 @@
+class FeeEstInput
+{
+public:
+    FeeEstInput(CBlockPolicyEstimator& estimator);
```
```
2018-09-20 05:13:44 cppcheck(pr=10443): [src/policy/fees_input.h:23]: (style) Class 'FeeEstInput' has a constructor with 1 argument that is not explicit.
```

practicalswift replied:
```
2018-09-22 22:51:38 cpplint(pr=10443): src/policy/fees_input.h:23:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,332 @@
+#include <policy/fees.h>
```
```
2018-09-22 22:51:38 cpplint(pr=10443): src/test/fee_est/fee_est.cpp:0:  No copyright message found.
```

practicalswift commented on src/policy/fees_input.h:
```diff
@@ -0,0 +1,55 @@
+
+    /** Process all the transactions that have been included in a block */
+    struct Block;
+    void processBlock(int height, const std::function<void(Block&)>& process_txs);
```
Isn't `height` guaranteed to be non-negative? Should be `unsigned int`?

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,222 @@
+    }
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        std::ifstream file(filename.string(), std::ifstream::binary);
```
The variable `file` shadows an already existing local variable. Please choose another name :-)

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,222 @@
+            int height = block["height"].get_int();
+            estimator.processBlock(height, [&](CBlockPolicyEstimator::Block& est_block) {
+                const auto& txs = value["txs"].getValues();
+                for (const UniValue& tx : txs) {
```
`tx` shadows a local variable. Please choose another name :-)

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,222 @@
+            std::vector<unsigned char> data = ParseHex(read.get_str());
+            uint16_t randv = 0;
+            GetRandBytes((unsigned char*)&randv, sizeof(randv));
+            fs::path filename = strprintf("fee_estimates.tmp.%04x", randv);
```
`filename` shadows another variable. Please use another name :-)

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,222 @@
+            CAutoFile(fsbridge::fopen(filename, "wb"), SER_DISK, CLIENT_VERSION)
+                .write((const char*)data.data(), data.size());
+            {
+                CAutoFile file(fsbridge::fopen(filename, "rb"), SER_DISK, CLIENT_VERSION);
```
Same here: shadows another local variable :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+            UniValue fee(UniValue::VOBJ);
+            fee.pushKV("mode", conservative ? "conservative" : "default");
+            fee.pushKV("blocks", int(target));
+            int64_t feeRate = estimator.estimateSmartFee(target, nullptr /* confTarget */, conservative).GetFeePerK();
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+        } else if (sample.size() < SAMPLE_TXS) {
+            sample.emplace_back(&entry);
+        } else {
+            int j = randint() % i;
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+        } else {
+            int j = randint() % i;
+            if (j < SAMPLE_TXS) {
+                sample[j] = &entry;
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+        std::vector<int64_t> feeRates;
+        feeRates.reserve(sample.size());
+        for (const auto& entry : sample) {
+            feeRates.emplace_back(CFeeRate(entry->second.fee, entry->second.size).GetFeePerK());
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+    UniValue txs(UniValue::VARR);
+    for (const auto& entry : sample) {
+        if (entry->second.actualBlocks > 0) {
+            int64_t feeRate = CFeeRate(entry->second.fee, entry->second.size).GetFeePerK();
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+{
+    if (*tx.first.begin() == 0) {
+        auto it = std::lower_bound(targets.begin(), targets.end(), nullptr, [&](unsigned int target, std::nullptr_t) {
+            CAmount estFee = estimator.estimateSmartFee(target, nullptr /* feeCalc */, false /* conservative */)
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+    if (*tx.first.begin() == 0) {
+        auto it = std::lower_bound(targets.begin(), targets.end(), nullptr, [&](unsigned int target, std::nullptr_t) {
+            CAmount estFee = estimator.estimateSmartFee(target, nullptr /* feeCalc */, false /* conservative */)
+                                 .GetFee(tx.second.size);
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+                                 .GetFee(tx.second.size);
+            return estFee <= 1 || tx.second.fee < estFee;
+        });
+        tx.second.expectedBlocks = it == targets.end() ? std::numeric_limits<int>::max() : *it;
```
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+}
+
+void PrintCross(const TxMap& txMap,
+    CBlockPolicyEstimator& estimator)
```
The unused parameter `estimator` can be removed?

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,333 @@
+        if (block.isObject()) {
+            int blockHeight = block["height"].get_int();
+            const auto& txs = value["txs"].getValues();
+            for (const UniValue& tx : txs) {
```
`tx` here shadows an already existing local variable. Please choose another name :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,330 @@
+    if (!sample.empty()) {
+        std::vector<CAmount> feeRates;
+        feeRates.reserve(sample.size());
+        for (const auto& entry : sample) {
```
Could use `std::transform`? :-)

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,330 @@
+            errsq += err * err;
+        }
+    }
+    printf("Non-test txs: %i\n", nonTestTxs);
```
Avoid `printf(…)` – it is not used in the rest of the project:

```
$ git grep "[^frns]printf" -- "*.cpp" "*.h" ":(exclude)src/leveldb/" ":(exclude)src/secp256k1/" ":(exclude)src/tinyformat.h" ":(exclude)src/univalue/"
$
```

practicalswift commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,330 @@
+                tx.pushKV("feeRate", feeRate);
+                tx.pushKV("blocks", entry->second.actualBlocks);
+                tx.pushKV("height", entry->second.height);
+                txs.push_back(std::move(tx));
```
This `move` is not needed?

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,227 @@
+            block->json_txs.push_back(std::move(tx));
+        } else {
+            UniValue value(UniValue::VOBJ);
+            value.pushKV("tx", std::move(tx));
```
`move` not needed?

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,227 @@
+        tx.pushKV("fee", fee);
+        tx.pushKV("size", uint64_t(size));
+        if (block) {
+            block->json_txs.push_back(std::move(tx));
```
`move` not needed?

practicalswift commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,227 @@
+            CAutoFile(fsbridge::fopen(data_filename, "wb"), SER_DISK, CLIENT_VERSION)
+                .write((const char*)data.data(), data.size());
+            {
+                CAutoFile data(fsbridge::fopen(data_filename, "rb"), SER_DISK, CLIENT_VERSION);
```
This `data` shadows the `data` on line 206 above :-)

ryanofsky commented:
Rebased 51163cee78f32fb6311bf620266c42fd115ba98f -> 8b9c3da3cf077051d6b1112ba4adbadd60096d57 ([pr/fee.16](https://github.com/ryanofsky/bitcoin/commits/pr/fee.16) -> [pr/fee.17](https://github.com/ryanofsky/bitcoin/commits/pr/fee.17)) due to conflict with #13341
Rebased 8b9c3da3cf077051d6b1112ba4adbadd60096d57 -> 4bcecdb666d1dbc90201dccbfee27a234f88c960 ([pr/fee.17](https://github.com/ryanofsky/bitcoin/commits/pr/fee.17) -> [pr/fee.18](https://github.com/ryanofsky/bitcoin/commits/pr/fee.18)) due to conflict with #13786
Updated 4bcecdb666d1dbc90201dccbfee27a234f88c960 -> aaa4b687c5fdeb2c75246e23077c984a0f133e81 ([pr/fee.18](https://github.com/ryanofsky/bitcoin/commits/pr/fee.18) -> [pr/fee.19](https://github.com/ryanofsky/bitcoin/commits/pr/fee.19)) fixing travis size_t univalue errors
Rebased aaa4b687c5fdeb2c75246e23077c984a0f133e81 -> a5faeb1cedd52b2886a1d5c320802874436fe2f6 ([pr/fee.19](https://github.com/ryanofsky/bitcoin/commits/pr/fee.19) -> [pr/fee.20](https://github.com/ryanofsky/bitcoin/commits/pr/fee.20)) to fix broken lint error caused by #13705
Updated a5faeb1cedd52b2886a1d5c320802874436fe2f6 -> 61ab4c17eebf9b73267c992fb664d603620f9e81 ([pr/fee.20](https://github.com/ryanofsky/bitcoin/commits/pr/fee.20) -> [pr/fee.21](https://github.com/ryanofsky/bitcoin/commits/pr/fee.21)) adding missing lock annotation.
Rebased 61ab4c17eebf9b73267c992fb664d603620f9e81 -> a9eb5cc5ceace3e567a0f7ba44c6350b2746436f ([pr/fee.21](https://github.com/ryanofsky/bitcoin/commits/pr/fee.21) -> [pr/fee.22](https://github.com/ryanofsky/bitcoin/commits/pr/fee.22)) due to conflict with #14100
Updated a9eb5cc5ceace3e567a0f7ba44c6350b2746436f -> 8cd1bb684bfec22fc54e091dc96c782a26847393 ([pr/fee.22](https://github.com/ryanofsky/bitcoin/commits/pr/fee.22) -> [pr/fee.23](https://github.com/ryanofsky/bitcoin/commits/pr/fee.23)) to fix `G_TRANSLATION_FUN` link error after #13961
Rebased 8cd1bb684bfec22fc54e091dc96c782a26847393 -> a60d26ac2bceea2c67ef34c8641705b16f656d07 ([pr/fee.23](https://github.com/ryanofsky/bitcoin/commits/pr/fee.23) -> [pr/fee.24](https://github.com/ryanofsky/bitcoin/commits/pr/fee.24)) due to conflict with #13311
Updated a60d26ac2bceea2c67ef34c8641705b16f656d07 -> 1e7ec287c927437b5a183bc807c49eba4a16d18a ([pr/fee.24](https://github.com/ryanofsky/bitcoin/commits/pr/fee.24) -> [pr/fee.25](https://github.com/ryanofsky/bitcoin/commits/pr/fee.25)) with suggested lint fixes
Updated 1e7ec287c927437b5a183bc807c49eba4a16d18a -> 56446378f5d0c2bdeb7a45fa7382ccd58c9b994c ([pr/fee.25](https://github.com/ryanofsky/bitcoin/commits/pr/fee.25) -> [pr/fee.26](https://github.com/ryanofsky/bitcoin/commits/pr/fee.26)) with more lint fixes
Updated 56446378f5d0c2bdeb7a45fa7382ccd58c9b994c -> 8d99281859540689b55285bd0acdf8c1529c4b37 ([pr/fee.26](https://github.com/ryanofsky/bitcoin/commits/pr/fee.26) -> [pr/fee.27](https://github.com/ryanofsky/bitcoin/commits/pr/fee.27)) with more lint fixes
Updated 8d99281859540689b55285bd0acdf8c1529c4b37 -> 68944a5120c651772fd7f3c028c446e38f9a3f57 ([pr/fee.27](https://github.com/ryanofsky/bitcoin/commits/pr/fee.27) -> [pr/fee.28](https://github.com/ryanofsky/bitcoin/commits/pr/fee.28)) with more lint fixes
Rebased 68944a5120c651772fd7f3c028c446e38f9a3f57 -> 50cd9b1cb7d03a81140d8a59698eba8d06b89612 ([pr/fee.28](https://github.com/ryanofsky/bitcoin/commits/pr/fee.28) -> [pr/fee.29](https://github.com/ryanofsky/bitcoin/commits/pr/fee.29)) due to conflict with #14555
Rebased 50cd9b1cb7d03a81140d8a59698eba8d06b89612 -> a0148157c57e91b299e6b450c7ee758f606488e6 ([pr/fee.29](https://github.com/ryanofsky/bitcoin/commits/pr/fee.29) -> [pr/fee.30](https://github.com/ryanofsky/bitcoin/commits/pr/fee.30)) due to conflict with #13128
Updated a0148157c57e91b299e6b450c7ee758f606488e6 -> 6f9c07f0a72ff0d95bfee9f37e4ae3dc9ed9c0e4 ([pr/fee.30](https://github.com/ryanofsky/bitcoin/commits/pr/fee.30) -> [pr/fee.31](https://github.com/ryanofsky/bitcoin/commits/pr/fee.31)) restructuring to work around clang lock annotation errors from #13128

Empact commented on src/policy/fees.cpp:
```diff
@@ -546,11 +545,13 @@ CBlockPolicyEstimator::~CBlockPolicyEstimator()
 {
 }
 
-void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, bool validFeeEstimate)
```
nit: this renaming is succinct, yes, but also less communicative.

Empact commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,330 @@
+
+using TxMap = std::map<uint256, TxData>;
+
+const char* const GRAPH_HTML = R"(<!DOCTYPE html>
```
nit: maybe load this from a separate file?

sipa commented:
Concept ACK, but it seems the HTML code (or even the conversion to HTML) could live outside the repo?

ryanofsky commented:
This PR has some overlap with @kallewoof's [mff](https://github.com/kallewoof/mff) utility. One difference is that this creates logs in [json-lines](http://jsonlines.org/) format and only logs mempool data needed to run fee estimation code reproducibly, while mff uses a custom format and logs additional information.

jonatack commented:
Discovered this PR thanks to DrahtBot's "needs rebase" notifications.

Concept ACK, interesting idea for dev tooling in the tests or contrib.

fanquake commented on src/policy/fees_input.cpp:
```diff
@@ -0,0 +1,222 @@
+    if (filename.empty()) {
+        log.reset();
+    } else {
+        log = MakeUnique<std::ofstream>(filename, std::ofstream::out | std::ofstream::app);
```
Please [use `std::make_unique` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/util/memory.h#L13).

ryanofsky commented:
Rebased 2c512ff0729f89aa47019009534d3958ec18f42b -> de8a2d80f74d33067a3b532368cca877fc65e7d2 ([`pr/fee.60`](https://github.com/ryanofsky/bitcoin/commits/pr/fee.60) -> [`pr/fee.61`](https://github.com/ryanofsky/bitcoin/commits/pr/fee.61), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/fee.60-rebase..pr/fee.61)) due to conflicts with #18766

aureleoules commented:
Concept ACK
CI is failing though.

aureleoules commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,331 @@
+void WriteGraph(const std::string& filename,
+    const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
+    CBlockPolicyEstimator& estimator)
```
538c780d078ecf400eb7b509f9fbcd176cdd290c
```suggestion
    const CBlockPolicyEstimator& estimator)
```

aureleoules commented on src/policy/fees.cpp:
```diff
@@ -527,8 +526,8 @@ bool CBlockPolicyEstimator::_removeTx(const uint256& hash, bool inBlock)
-CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath)
-    : m_estimation_filepath{estimation_filepath}, nBestSeenHeight{0}, firstRecordedHeight{0}, historicalFirst{0}, historicalBest{0}, trackedTxs{0}, untrackedTxs{0}
+CBlockPolicyEstimator::CBlockPolicyEstimator()
+    : nBestSeenHeight{0}, firstRecordedHeight{0}, historicalFirst{0}, historicalBest{0}, trackedTxs{0}, untrackedTxs{0}
```
55016bfa39f496c08ea0c18260a3b6abbdb23550
nit: use class initialization

aureleoules commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,331 @@
+    return true;
+}
+
+void PrintCross(const TxMap& txMap, CBlockPolicyEstimator&)
```
538c780d078ecf400eb7b509f9fbcd176cdd290c
```suggestion
void PrintCross(const TxMap& txMap)
```

aureleoules commented on src/test/fee_est/fee_est.cpp:
```diff
@@ -0,0 +1,331 @@
+    file << strprintf(GRAPH_HTML, txs.write(), fees.write());
+}
+
+bool UpdateCross(TxMap::value_type& tx, const std::vector<unsigned int>& targets, CBlockPolicyEstimator& estimator)
```
538c780d078ecf400eb7b509f9fbcd176cdd290c
```suggestion
bool UpdateCross(TxMap::value_type& tx, const std::vector<unsigned int>& targets, const CBlockPolicyEstimator& estimator)
```

josibake commented:
Haven't dug into the code, but I'm leaning toward Concept ACK, Approach NACK

As someone who is planning to do fee estimation research in the near future, I don't think I'd use something like this the way it's currently written. I would much prefer something that logs Bitcoin Core's live _fee rate_ estimations, or some option that would allow me to tell Bitcoin Core it's at a certain chain height, feed it some mempool data, and then ask it for fee estimations.

I also agree that the tool for analyzing the data shouldn't be in this repo. Much better if we can figure out what data is missing for people who want to do fee estimation analysis and provide an option for logging that data.

Perhaps a good next step would be to talk to people who are currently doing fee estimation work and see what data would help with their projects/PRs? @ClaraShk @darosior @Xekyo ?

ryanofsky commented:
@josie thanks for looking at this. I don't think I understand what the idea behind an approach NACK would be based on what you are saying, though.

It sounds more like you're asking for an additional feature which would be easy to implement on top of the first commit, not pointing to a drawback of the approach. If you want to gather live fee estimates, the first commit outputs a structured [jsonl](https://jsonlines.org/) log of fee estimation events, so it would be easy to add the actual fee estimates to this log, or add internal metrics used to produce the fee estimates, or add any other data points that might be interesting. It is also possible to take the existing log files this PR already generates, and feed them to the fee estimation algorithm by modifying the tool in the second commit to produce live fee estimates or a history of estimates over time.

If you just think I should drop the second commit, or trim it down, I think that would be reasonable. The point of the second commit is mainly to provide example code that demonstrates how to read the JSONL log file and how to call the fee estimator API. The idea is to avoid needing reinvent the wheel each time you want to monitor fee estimation behavior, or reproduce a bug, or determine the effects of a change.

> I would much prefer [...] some option that would allow me to tell Bitcoin Core it's at a certain chain height, feed it some mempool data, and then ask it for fee estimations.

This is actually what the second commit provides. The .jsonl log contains chain and mempool data. You can feed it to the test program in the second commit, and it calls fee estimation code to produce fee estimates. It can produce a graph of the fee estimates, and calculate how accurate they were by looking at all the transactions that got confirmed and outputting the mean squared error of how many blocks it took the transactions to be confirmed vs how many blocks the fee estimation algorithm predicted it would take to get them confirmed based on their feerate.

This PR is a few years old, so I would probably implement the second commit a little differently today. Instead of adding a standalone c++ executable that calls fee estimator code and produces graphs and metrics, I might want to just add a python module that exposes the `CBlockPolicyEstimator` class, and generate the graphs and metrics from a python notebook.

josibake commented:
Thanks for the thorough response @ryanofsky. I realize I didn't explain myself very well in my initial comment.

> It sounds more like you're asking for an additional feature which would be easy to implement on top of the first commit, not pointing to a drawback of the approach

Since we already record mempool events in the debug log file (this is what @jamesob and I are using to log mempool events in [bmon](https://github.com/chaincodelabs/bmon)), it feels a bit strange to have special code for logging transactions again here for fee estimation. It's possible I'm not understanding something here, though: is there some special information in the log you are writing that we can't get from the debug log file?

> This is actually what the second commit provides. The .jsonl log contains chain and mempool data. You can feed it to the test program in the second commit, and it calls fee estimation code to produce fee estimates

I think what I was envisioning was even simpler: have a flag to make Bitcoin Core log `fee_rate` estimates per block and then have some way to walk the node back to some point in the past, have it log `fee_rate` estimates (perhaps with a code change) and then leave it up to the individual on how to analyze the diff between those logs. Perhaps this is easier said than done since the fee estimation code in Bitcoin Core today seems to work on a per transaction basis?

ryanofsky commented:
> is there some special information in the log you are writing that we can't get from the debug log file?

I think so. The json log captures all inputs to fee estimator to used update its internal state. It includes not just mempool events but also information about which transactions are included in blocks, flush events, and contents of `fee_estimates.dat` blobs that are loaded. It makes fee estimation code reproducible, so if you make a change to the fee estimation code or parameters, you can evaluate the effects of the change with real world inputs offline. [^1]

[^1]: I also think in practice parsing a json log file is cleaner and more straightforward than parsing a debug.log file. Compare
[readLog](https://github.com/ryanofsky/bitcoin/blob/26240e1a5f320e8a9da71cf4b7e6313a9b494aff/src/policy/fees_input.cpp#L162-L228) vs [logparse](https://github.com/chaincodelabs/bmon/blob/26032dbec86f14044c9af0af0e06528750ad6fc8/bmon/logparse.py). But I recognize there could be other advantages to using debug.log files, especially since debug.log files are omnipresent and already exist.

> I think what I was envisioning was even simpler: have a flag to make Bitcoin Core log fee_rate estimates per block and then have some way to walk the node back to some point in the past, have it log fee_rate estimates (perhaps with a code change) and then leave it up to the individual on how to analyze the diff between those logs. Perhaps this is easier said than done since the fee estimation code in Bitcoin Core today seems to work on a per transaction basis?

I think I don't actually understand the proposal or don't understand the intended use-case. Use-cases for PR this are to:

- Be able to make changes to the fee estimation algorithm and be able to see how the changes affect fee estimates in a quick feedback loop
- Be able to evaluate quality of fee estimates. The example program compares how quickly fee estimator expects transactions get to get confirmed (based on feerate) vs how quickly they actually get confirmed. This could be one metric for comparing quality of different fee estimation algorithms, or seeing if one algorithm starts to perform better or worse over time. It should also be possible to implement other metrics besides this one and do similar things.

Is your proposal basically to save a `fee_estimates.dat` file each time a block is connected and then use those files for something? I'm guess I'm not sure how you would use those files to improve fee estimates. This PR does capture a superset of that information, too. For example, it would be possible to generate those files offline from the json log file produced in this PR by adding a few lines to the [`filter`](https://github.com/ryanofsky/bitcoin/blob/26240e1a5f320e8a9da71cf4b7e6313a9b494aff/src/test/fee_est/fee_est.cpp#L285) callback:

```c++
const UniValue& block = value["block"];
if (block.isObject()) {
    AutoFile file{fsbridge::fopen(strprintf("fee_estimates-%07d.dat", block["height"].getInt<int>());, "wb")};
    estimator.Write(file);
}
```

Another advantage of this change is that it decouples `CBlockPolicyEstimator` from the mempool, so it's possible to call it offline. The external API changes from:

https://github.com/bitcoin/bitcoin/blob/7b45d171f549595a831489827c28e8493f36c00c/src/policy/fees.h#L189-L200

to:

https://github.com/bitcoin/bitcoin/blob/26240e1a5f320e8a9da71cf4b7e6313a9b494aff/src/policy/fees.h#L191-L201

There are probably lots of changes that could be made to this PR to improve it, or trim it down if it is doing too much, but in general I think takes a good first step of pulling things out of `CBlockPolicyEstimator` that don't belong there (mempool access, filesystem access), and then builds some useful features on top of that.

ryanofsky commented:
Going to close this because I think @josibake is working on a different approach which will try replace current fee estimator, instead of just instrumenting it like this PR.

There's also another PR #28368 which decouples the fee estimator from the mempool, and might make it easier to replay fee estimation events like the changes in this PR, even though it only provides an API, not a CLI tool.
